src/front/wgsl.rs

```diff
@@ -548,7 +548,7 @@ impl Parser {
             other => return Err(Error::Unexpected(other)),
         };
         self.scopes.pop();
-        Ok(ctx.expressions.append(expression))
+        Ok(ctx.expressions.fetch_or_append(expression))
```

Comment:
Not sure this is a good idea. There is a cost to `fetch_or_append` proportional to the number of elements, and it's not optimized yet. There may be many many expressions, so checking for duplication on each one may be expensive.

Reply:
Yes that is true, but then there will be duplicate expressions in this Arena. If I use this simple WGSL:
```
fn main() -> void {
  var a : vec4<f32> = vec4<f32>(1.0, 0.0, 0.0, 1.0);
  return;
}
entry_point fragment as "main" = main;
```

With only `append` the IR looks like this:
```
expressions: Arena {
	data: [
		Constant(
			Handle(1),
		),
		Constant(
			Handle(2),
		),
		Constant(
			Handle(2),
		),
		Constant(
			Handle(1),
		),
		Compose {
			ty: Handle(1),
			components: [
				Handle(1),
				Handle(2),
				Handle(3),
				Handle(4),
			],
		},
		LocalVariable(
			Handle(1),
		),
	],
},
```
As you can see you would get double constant expressions refering to the same Handle.

While using it with `fetch_or_append`, would result in this:
```
expressions: Arena {
	data: [
		Constant(
			Handle(1),
		),
		Constant(
			Handle(2),
		),
		Compose {
			ty: Handle(1),
			components: [
				Handle(1),
				Handle(2),
				Handle(2),
				Handle(1),
			],
		},
		LocalVariable(
			Handle(1),
		),
	],
},
```
This is only a single line of WGSL and in the future it would also exponentially increase the amount of expressions in the IR.

Now I understand the performance issue, and it will be a trade-off:
1 ) Cluttering the IR with duplicate expressions and let the back-end take care of the duplicates by not including them. There still needs to be a check in the back-end for duplicates. I am not sure if that is less, even or more expensive.
2 ) Checking in the front-end and prevent from ever inserting duplicates in the IR

Reply:
Thank you for the detailed comparison, it's very helpful for the discussion!
It comes down to what we call an expression in the IR. If it's an expression in the AST sense, then it makes total sense to deduplicate them (at least, if it costed us nothing, but I think it does).
If it's an expression in the SPIR-V sense (i.e. an SSA value), then we shouldn't be deduplicating at least some of the expressions (different loads from a variable, for example).

It's not clear to me where to draw the line at this point, but it's a safe bet to not deduplicate the expressions at least today. Let's keep it, at least until we settle on a stronger meaning of an "expression", as well as have a faster way (lookup hashmap?) to deduplicate them (current one is too naive).

Reply:
Agreed, I think trying to "deduplicate" the expressions here are as of yet an unknown cost. 

Lets revert this one too and only keep the constants as `fetch_or_append`. That is my use-case for the SPIR-V back-end at the moment: it is generating a lot of OpTypeInt's and OpTypeFloats with the same value.

In the meantime, while Naga is maturing we can always resume this discussion. :)